### DISCUSSION TITLE
Get not allowing data files json

### DISCUSSION THREAD

**Discussion by tridentsx** [encode/httpx#1546] [Potential Issue]
What if I want to test the behavior of a http server when receiving a body with get request. How should I do it ? This is no problem in requests library, they do support these parameters. I also need http/2 support.

From the HTTP 1.1 2014 Spec:
=============================
A payload within a GET request message has no defined semantics; sending a payload body on a GET request might cause some existing implementations to reject the request.

**Comment by StephenBrown2** [comment]
Hi! Per the docs, which mention it under [Requests Compatibility](https://www.python-httpx.org/compatibility/) > [Request body on HTTP methods](https://www.python-httpx.org/compatibility/#request-body-on-http-methods):

> If you really do need to send request data using these http methods you should use the generic `.request` function instead.

Documented at the top of: https://www.python-httpx.org/api/

Example:
```pycon
>>> import httpx
>>> response = httpx.request('GET', 'https://httpbin.org/anything', json={'key': 'value'})
>>> response
<Response [200 OK]>
>>> response.json()
{'args': {},
 'data': '{"key": "value"}',
 'files': {},
 'form': {},
 'headers': {'Accept': '*/*',
  'Accept-Encoding': 'gzip, deflate',
  'Content-Length': '16',
  'Content-Type': 'application/json',
  'Host': 'httpbin.org',
  'User-Agent': 'python-httpx/0.17.0',
  'X-Amzn-Trace-Id': 'Root=1-60622ae6-1d1cff9879dc8c341bd8fcdc'},
 'json': {'key': 'value'},
 'method': 'GET',
 'origin': '72.32.180.176',
 'url': 'https://httpbin.org/anything'}
```
You can also create your own Request instances as documented here: https://www.python-httpx.org/advanced/#request-instances

**Reply by tridentsx** [reply]
I need it to work with client/session also and will use it from a robot library that originally used requests, dont want to change to much code, not sure I understand the reasoning of removing valid http requests from the interface, regardless if body is ignored by the http server. httpx is not implementing a server its implementing a client wich should be able to send any valid http request

**Reply by StephenBrown2** [reply]
Also from the docs,
> The HTTP `GET`, `DELETE`, `HEAD`, and `OPTIONS` methods are specified as not supporting a request body.

so a `GET` with a body is not technically a valid http request. Several decisions have been made to break with requests compatibility to reduce incorrect usage, as well as keep things clear and concise. `httpx` does not advertise itself as a 100% drop-in replacement for every bit of requests, but "broadly compatible".

You can see the discussion that surrounded this decision here: https://github.com/encode/httpx/issues/395#issuecomment-535864986

However, to answer your question, `Client` objects do also have a `request` function:
```py
with httpx.Client() as client:
    response = client.request("GET", "https://httpbin.org/anything", json={"key": "value"})
```

**Comment by tridentsx** [comment]
A parameter having an undefined usage or semantics is not the same as it not being valid, from RFC7231

   A payload within a GET request message has no defined semantics;
   sending a payload body on a GET request might cause some existing
   implementations to reject the request. 


From RFC7230 its very clear body with get is very much valid, where the standard explicitly points out one case where body is not valid and its in a response to a HEAD request.

3.3.  Message Body

   The message body (if any) of an HTTP message is used to carry the
   payload body of that request or response.  The message body is
   identical to the payload body unless a transfer coding has been
   applied, as described in Section 3.3.1.

     message-body = *OCTET

   The rules for when a message body is allowed in a message differ for
   requests and responses.

   The presence of a message body in a request is signaled by a
   Content-Length or Transfer-Encoding header field.  Request message
   framing is independent of method semantics, even if the method does
   not define any use for a message body.

   The presence of a message body in a response depends on both the
   request method to which it is responding and the response status code
   (Section 3.1.2).  Responses to the HEAD request method (Section 4.3.2
   of [RFC7231]) never include a message body because the associated
   response header fields (e.g., Transfer-Encoding, Content-Length,
   etc.), if present, indicate only what their values would have been if
   the request method had been GET (Section 4.3.1 of [RFC7231]). 2xx
   (Successful) responses to a CONNECT request method (Section 4.3.6 of
   [RFC7231]) switch to tunnel mode instead of having a message body.
   All 1xx (Informational), 204 (No Content), and 304 (Not Modified)
   responses do not include a message body.  All other responses do
   include a message body, although the body might be of zero length.

**Reply by lovelydinosaur** [reply]
We're not preventing anyone from doing *this slightly weird thing that you probably didn't mean to do*, we're just making sure that if you want to do *this slightly weird thing that you probably didn't mean to do* then you need to type a few more characters.

`httpx.request("GET", ...)` instead of `httpx.get(...)`.

In exchange for that, we help guide the majority of users along the happy path of actually meaningful HTTP usage.
